### PR TITLE
Solution Entities

### PR DESCRIPTION
**Pull request by TemporalOroboros**
## About the PR

Solutions that are associated with an entity by a `SolutionContainerComponent` are now entities with all of the attendant API changes that requires.

## Why / Balance

In the future storing the progress of chemical reactions or the effects of reagents can be done on a per-solution basis instead of a per-container basis.

For example, a beaker made from uranium class that also contains radium would be radioactive/luminous from both the material it is made of and the reagents it contains. As a result, the system managing this radiation would need to contend with any innate radiation the beaker would have, the materials of the beaker, and the solutions in the beaker without any of them clobbering the others.

As a bonus, this gets us slightly closer to unifying materials and reagents into the same system.

## Technical details

Solutions stored in entities like beakers, tanks, or bloodstreams, are now stored as entities in containers with a `SolutionComponent` to track the state of the solution. Which solutions are stored within a given entity is tracked via a `SolutionContainerManagerComponent` on the containing entity. The actual solution entities are spawned during mapinit according to raw solution data original stored on the `SolutionContainerManagerComponent`.

Manipulating the state of given solutions and accessing the solutions associated with a given container is done through the `SharedSolutionContainerSystem` and its descendants.

## Media

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

1. The `SolutionContainerSystem` has been converted into the abstract `SharedSolutionContainerSystem` with concrete `SolutionContainerSystem`s in the client and server namespaces.<br/>If you used it please switch the used namespace to refer to the most relevant version.
2. All of the methods on `SolutionContainerSystem` that took a `Solution` and the `EntityUid` of the containing entity have been modified to take an `Entity<SolutionComponent>` instead.<br/>If you used any of them, please supply the given solution entity (accessible through `TryGetSolution`) as is now required.
3. The `TryGetSolution` method has had a full parameter overhaul.
  - It now takes an `Entity<SolutionContainerManagerComponent?>` instead of a separate `EntityUid?` and `SolutionContainerManagerComponent?`.<br/>Compensating for this will require packing any passed component reference into the first argument as a tuple with the uid. Note that this means that the passed uid is no longer nullable and may require earlier null checks if you were relying on that behavior.
  - The name of the associated solution to fetch is now nullable and passing `null` will attempt to interpret the solution container entity itself as a solution entity.
  - Now has a second out parameter emitting an `Entity<SolutionComponent>` alongside the parameter directly returning the `Solution` reference.<br/>You will need to add an additional out parameter and may need to pass the emitted solution entity to any relevant future methods.
4. The `EnsureSolution` methods have been moved to the server version of `SolutionContainerSystem`. If you used them in the shared namespace please move the relevant code to the server.
5. The `Solutions` field on `SolutionContainerManagerComponent` no longer tracks the state of the solutions associated with the entity past mapinit.</br>If you were using it to access the solutions associated with the entity use the `TryGetSolution` or `EnumerateSolutions(Entity<...>)` on the associated system.

### CODE REVIEW

**Comment by TemporalOroboros** [comment]
It is ~midnight and I need to sleep. I will do the tech details and breaking changes tomorrow.

File: Content.Shared/Chemistry/Containers/Components/ContainerSolutionComponent.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+using Content.Shared.Chemistry.Containers.EntitySystems;
+using Robust.Shared.GameStates;
+using Robust.Shared.Serialization;
+
+namespace Content.Shared.Chemistry.Containers.Components;
+
+/// <summary>
+/// 
+/// </summary>
```
Comment (ElectroJr) [inline Content.Shared/Chemistry/Containers/Components/ContainerSolutionComponent.cs:9]: empty summary

File: Content.Shared/Chemistry/Containers/Components/ContainerSolutionComponent.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    [DataField]
+    public EntityUid Container;
+
+    [DataField]
+    public string Name = default!;
+}
+
+
+[Serializable, NetSerializable]
+public sealed partial class ContainerSolutionState : ComponentState
```
Comment (ElectroJr) [inline Content.Shared/Chemistry/Containers/Components/ContainerSolutionComponent.cs:23]: auto comp states

File: Content.Shared/Chemistry/Solutions/Solution.cs [hunk 3]
```diff
@@ -0,0 +1,880 @@
...
+
+namespace Content.Shared.Chemistry.Solutions;
+
+/// <summary>
+///     A solution of reagents.
+/// </summary>
+[Serializable, NetSerializable]
+[DataDefinition]
+public sealed partial class Solution : IEnumerable<ReagentQuantity>, ISerializationHooks
+{
```
Comment (ElectroJr) [inline Content.Shared/Chemistry/Solutions/Solution.cs:18]: All of the namespaces changes should probably be a separate PR because it makes reviewing much harder.

File: Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+
+    private Entity<SolutionComponent> SpawnSolutionUninitialized(EntityUid container, string name, Solution prototype)
+    {
+        var coords = new EntityCoordinates(container, Vector2.Zero);
+        var uid = EntityManager.CreateEntityUninitialized(null, coords, null);
+        if (!TryComp(uid, out SolutionComponent? comp))
+        {
+            var solution = prototype.Clone();
```
Comment (ElectroJr) [inline Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs:143]: This is currently only ever called with a new solution, so this is currently always an unnecessary clone.
Seeing as this is a private method anyways, it should be fine to just use the solution passed in without any cloning.

File: Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        if (!TryComp(solutionId, out ContainerSolutionComponent? relation))
+        {
+            relation = new ContainerSolutionComponent() { Container = uid, Name = name };
+            AddComp(solutionId, relation);
+        }
+        else
+        {
+            (relation.Container, relation.Name) = (uid, name);
+            Dirty(solutionId, relation);
+        }
```
Comment (ElectroJr) [inline Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs:105]: This already gets set by `SpawnSolutionUninitialized()`, which should be the only way that a solution entity is ever spawned. So for the sake of performance it'd be better to just remove this or make it an assert in a debug block.

File: Content.Server/Fluids/EntitySystems/DrainSystem.cs [hunk 6]
```diff
@@ -206,8 +205,8 @@ public override void Update(float frameTime)
     private void OnExamined(EntityUid uid, DrainComponent component, ExaminedEvent args)
     {
         if (!args.IsInDetailsRange ||
-            !HasComp<SolutionContainerManagerComponent>(uid) ||
-            !_solutionSystem.TryGetSolution(uid, DrainComponent.SolutionName, out var drainSolution))
+            !HasComp<SolutionContainerComponent>(uid) ||
+            !_solutionContainerSystem.TryGetSolution(uid, DrainComponent.SolutionName, out _, out var drainSolution))
```
Comment (ElectroJr) [inline Content.Server/Fluids/EntitySystems/DrainSystem.cs:209]: I know its not your code, but this hascomp is redundant.

File: Content.Server/Chemistry/Containers/Components/SolutionContainerManagerComponent.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+using Content.Shared.Chemistry.Solutions;
+
+namespace Content.Server.Chemistry.Containers.Components;
+
+/// <summary>
+/// Component used to spawn solution entities on <see cref="MapInitEvent"/>.
+/// Replaces itself with a <see cref="SolutionContainerComponent"/> during map initialization.
+/// </summary>
+[RegisterComponent]
+[Access(typeof(SolutionContainerSystem))]
```
Comment (ElectroJr) [inline Content.Server/Chemistry/Containers/Components/SolutionContainerManagerComponent.cs:12]:
```suggestion
[Access(typeof(SolutionContainerSystem), Other = AccessPermissions.None)]
```
to avoid people accidentally using the wrong component.

File: Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+    private void OnMapInit(EntityUid uid, SolutionContainerManagerComponent comp, MapInitEvent args)
+    {
+        if (comp.Solutions is { Count: > 0 } prototypes)
+        {
+            foreach (var (name, prototype) in prototypes)
+            {
+                var solution = EnsureSolutionEntity(uid, name, prototype.MaxVolume, out _);
+                SolutionSystem.AddSolution(solution, prototype);
+            }
```
Comment (ElectroJr) [inline Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs:171]: I'd say just re-use the existing solution object instead of creating a new one and adding its contents, because solution addition isn't all that fast and this migh slow down map initialization.

File: Content.Shared/Chemistry/Containers/Components/SolutionContainerComponent.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    /// The default amount of space that will be allocated for solutions in solution containers.
+    /// Most solution containers will only contain 1-2 solutions.
+    /// </summary>
+    public const int DefaultCapacity = 2;
+
+    /// <summary>
+    /// A map of solution names to solution entities attached to the owning entity.
+    /// </summary>
+    [ViewVariables]
+    public Dictionary<string, ContainerSlot> Solutions = new(DefaultCapacity);
```
Comment (ElectroJr) [inline Content.Shared/Chemistry/Containers/Components/SolutionContainerComponent.cs:30]: This component does need a data field to store the container ids, along some way to populate this field on init, so that you can save & load post map-init entities

File: Content.Server/Chemistry/Containers/Components/SolutionContainerManagerComponent.cs (hunk 7 above)
Comment (ElectroJr) [reply Content.Server/Chemistry/Containers/Components/SolutionContainerManagerComponent.cs:12]: Actually, why aren't the two components just the same component? you can just put the access restriction on the "prototype" field. This would also significantly reduce the number of changes.

**Review by ElectroJr — CHANGES_REQUESTED** [review]
Not a thorough review, just some stuff I saw while skimming through.

File: Content.Server/Chemistry/Containers/Components/SolutionContainerManagerComponent.cs (hunk 7 above)
Comment (TemporalOroboros) [reply Content.Server/Chemistry/Containers/Components/SolutionContainerManagerComponent.cs:12]: It's because I was assuming it would be better for perf to deallocate the prototype dictionary when we're done with it so we don't have ~twice as much memory persistently allocated for each solution container.

File: Content.Shared/Chemistry/Containers/Components/ContainerSolutionComponent.cs (hunk 2 above)
Comment (TemporalOroboros) [reply Content.Shared/Chemistry/Containers/Components/ContainerSolutionComponent.cs:23]: IIRC auto comp states can't handle the EntityUid because it's not netserializable.

File: Content.Shared/Chemistry/Containers/Components/ContainerSolutionComponent.cs (hunk 1 above)
Comment (TemporalOroboros) [reply Content.Shared/Chemistry/Containers/Components/ContainerSolutionComponent.cs]: Should be fixed now.

File: Content.Shared/Chemistry/Solutions/Solution.cs (hunk 3 above)
Comment (TemporalOroboros) [reply Content.Shared/Chemistry/Solutions/Solution.cs]: Moved the namespace changes back.

File: Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs (hunk 4 above)
Comment (TemporalOroboros) [reply Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs]: Done.

File: Content.Server/Chemistry/Containers/Components/SolutionContainerManagerComponent.cs (hunk 7 above)
Comment (TemporalOroboros) [reply Content.Server/Chemistry/Containers/Components/SolutionContainerManagerComponent.cs]: Made them the name comp

File: Content.Shared/Chemistry/Containers/Components/SolutionContainerComponent.cs (hunk 9 above)
Comment (TemporalOroboros) [reply Content.Shared/Chemistry/Containers/Components/SolutionContainerComponent.cs]: Done. Container slots are now accessed through `ContainerSystem.TryGetContainer`

**Comment by ElectroJr** [comment]
I should have time this weekend to do a full review

File: Content.Shared/Chemistry/Containers/Components/ContainerSolutionComponent.cs (hunk 2 above)
Comment (ElectroJr) [reply Content.Shared/Chemistry/Containers/Components/ContainerSolutionComponent.cs]: It automatically converts them to `NetEntity` and back

**Review by ElectroJr — CHANGES_REQUESTED** [review]
There are a couple of minor things, and two big-ish things. Other than that it seems fine.
Firstly, the various TryGet functions should return a nullable Entity<T> to avoid null-reference exceptions.

The other thing is that there should be a caching try-get method for the sake of performance, in order to avoid unnecessarily fetching the `SolutionContainerManagerComponent` and `ContainerManagerComponent` anytime you try to fetch a solution.

I.e., all the performance sensitive components/systems should use something like:


```cs
public sealed class LungComponent
{
    public string SolutionId;
    public Entity<SolutionComponent>? Solution;
}

public sealed class LungSystem
{
    public void UpdateLung(EntityUid uid, LungComponent lung)
    {
        if (!_solution.Resolve(uid, lung.SolutionId, ref lung.Solution))
            return;
        // do lung stuff
    }
}

public sealed class SolutionSystem
{
    public bool Resolve(Entity<SolutionContainerComponent?> container, string name, [NotNullWhen(true)]ref Entity<SolutionComponent>? solution, bool logError = true)
    {
        if (solution != null)
            return true;
        // try to fetch and set the solution.
    }

    public void EnsureSolution(Entity<SolutionContainerComponent?> container, string name, [NotNull] ref Entity<SolutionComponent>? solution)
    {
        if (solution != null)
            return true;
        // get or create a solution
    }
}
```

Not **every** system/component would need to cache it, but it'd be a good precedent to set even for unimportant systems. The most important systems are those that automatically update solutions on tick updates. I.e.: Puddle, Lung, Metabolizer, Stomach, Bloodstream, PlantHolder, and SolutionRegeneration. I'd say look at all the uses of the solution name fields for those components and cache the solution.

If thats done then I think it should be good, and shouldn't even hurt performance, aside from PVS having to deal with more entities.

File: Content.Server/Body/Systems/BloodstreamSystem.cs [hunk 10]
```diff
@@ -81,6 +79,16 @@ private void OnReactionAttempt(EntityUid uid, BloodstreamComponent component, Re
...
+    private void OnReactionAttempt(EntityUid uid, BloodstreamComponent component, ref SolutionRelayEvent<ReactionAttemptEvent> args)
+    {
+        if (args.Name != BloodstreamComponent.DefaultBloodSolutionName
+            && args.Name != BloodstreamComponent.DefaultChemicalsSolutionName
+            && args.Name != BloodstreamComponent.DefaultBloodTemporarySolutionName)
+            return;
+
+        OnReactionAttempt(uid, component, ref args.Event);
+    }
+
```
[the remark this answers is not in the document]
Comment (TemporalOroboros) [reply Content.Server/Body/Systems/BloodstreamSystem.cs:91]: It only blocks specific types of reactions (those that produce entities/smoke/foam) instead of all reactions.

File: Content.Shared/Chemistry/Reaction/ChemicalReactionSystem.cs [hunk 11]
```diff
@@ -307,15 +312,10 @@ public void FullyReactSolution(Solution solution, EntityUid owner, FixedPoint2 m
     /// <reamrks>
     ///     Some solution containers (e.g., bloodstream, smoke, foam) use this to block certain reactions from occurring.
     /// </reamrks>
-    public sealed class ReactionAttemptEvent : CancellableEntityEventArgs
+    public partial record struct ReactionAttemptEvent(ReactionPrototype Reaction, Entity<SolutionComponent> Solution)
```
[the remark this answers is not in the document]
Comment (TemporalOroboros) [reply Content.Shared/Chemistry/Reaction/ChemicalReactionSystem.cs:315]: Since DataDefinitions must be partial and most of the things I interact with are either DataDefinition or EntitySystem I have started to make everything I implement partial on reflex.

**Comment by TemporalOroboros** [comment]
Question about the caching. Should I just be caching the entityuid since storing a component on a component seems somewhat bad? Also, it will probably be incompatible with struct components seeing as that would make `Entity<TComp, ...>` a ref struct, which cannot be stored.

**Comment by ElectroJr** [comment]
> Should I set up an event handler to invalidate the cache if the solution entity gets deleted? I'm assuming yes for VV purposes.

It should be fine as long as you only cache solutions that shouldn't ever be deleted. I.e., a lung component without the lung solution is non-functional, it should only be deleted if the lung component itself is also removed. And in the event that something goes wrong and it does get deleted, it should be fine seeing as the field should be nullable & used with the `Resolve()` method anyways. 

> Should I just be caching the entityuid since storing a component on a component seems somewhat bad?

Struct components are probably still a long ways away, and not every component will need to be converted, so IMO its worth reducing the number of component lookups, especially if arch makes single component lookups slower.
And this is the sort of thing where it'd be pretty trivial to change it back to an EntityUid in future if we need to.

**Comment by Jezithyr** [comment]
Is there anything I can do to help out with getting this PR merged faster? As it stands this is holding up medical refactor since I need to do some work on metabolism/bloodstream and Solution Entities is a major change that has implications for both systems.
If you're busy I can make the suggested fixes and resolve conflicts so that we can get this merged asap.

**Comment by TemporalOroboros** [comment]
I should be able to get back to this today/tomorrow and can at least do the merge conflicts then. Do you know what bits of the reviews I still need to do? IIRC I did all of the reviews besides completely disabling bloodstream solutions (which I don't think is wanted?)

**Comment by AJCM-git** [comment]
> I should be able to get back to this today/tomorrow and can at least do the merge conflicts then. Do you know what bits of the reviews I still need to do? IIRC I did all of the reviews besides completely disabling bloodstream solutions (which I don't think is wanted?)

Did you do the caching? I added the tag cuz I didnt see a commit after your questions so I just assumed that still needed doing, if it's done this can prob get merged after solving conflicts

**Comment by TemporalOroboros** [comment]
> > I should be able to get back to this today/tomorrow and can at least do the merge conflicts then. Do you know what bits of the reviews I still need to do? IIRC I did all of the reviews besides completely disabling bloodstream solutions (which I don't think is wanted?)
> 
> Did you do the caching? I added the tag cuz I didnt see a commit after your questions so I just assumed that still needed doing, if it's done this can prob get merged after solving conflicts

Caching was done in e7353f4 for a couple most of the systems pointed out + a few more.
IIRC I left out metabolized because that one can have the solution change underneath it if the organ gets moved.

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Looks good, just some details.
Update the event handlers you touched to the Entity<TComp> pattern, only looked at SharedSolutionContainerSystem and SolutionContainerSystem but there may be others. If the reason you didnt do it was cuz of build errors, just  adding a ref to the events (even if they are value types) should do the trick.

Also try to tweak your formatter and namespace optimizer and resave your files since some of em have unused/wrongly-ordered `using`'s, like moving System.Linq at the bottom instead of the top (very minor detail, not really required)

File: Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerSystem.Capabilities.cs
[the remark this answers is not in the document]
Comment (TemporalOroboros) [reply Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerSystem.Capabilities.cs:1]: Converted to explicit regions.

File: Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerSystem.Relays.cs [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        {
+            Handled = args.Handled,
+        };
+
+        RaiseLocalEvent(comp.Container, ref relayEv);
+        args.Handled = relayEv.Handled;
+    }
+
+
+
```
[the remark this answers is not in the document]
Comment (TemporalOroboros) [reply Content.Shared/Chemistry/EntitySystems/SharedSolutionContainerSystem.Relays.cs:104]: Converted to explicit regions.

File: Content.Shared/Chemistry/Components/SolutionManager/ContainedSolutionComponent.cs [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    /// The entity that the solution is contained in.
+    /// </summary>
+    [DataField(required: true), AutoNetworkedField]
+    public EntityUid Container;
+
+    /// <summary>
+    /// The name/key of the container the solution is located in.
+    /// </summary>
+    [DataField(required: true), AutoNetworkedField]
+    public string Name = default!;
```
[the remark this answers is not in the document]
Comment (TemporalOroboros) [reply Content.Shared/Chemistry/Components/SolutionManager/ContainedSolutionComponent.cs:30]: Renamed to ContainerName

**Review by AJCM-git — APPROVED** [review]
